graphs/graph_list.py

```diff
@@ -14,137 +28,548 @@
+                undirected_graph, directed_graph, edge
+            )
+        self.assertFalse(undirected_graph.directed)
+        self.assertTrue(directed_graph.directed)
+
+    def test_contains_vertex(self):
+        random_vertices: list[int] = random.sample(range(101), 20)
+
+        # Build graphs WITHOUT edges
+        undirected_graph = GraphAdjacencyList(
+            vertices=random_vertices, edges=[], directed=False
+        )
+        directed_graph = GraphAdjacencyList(
+            vertices=random_vertices, edges=[], directed=True
+        )
+
+        # Test contains_vertex
+        for num in range(101):
+            self.assertEqual(
+                num in random_vertices, undirected_graph.contains_vertex(num)
+            )
+            self.assertEqual(
+                num in random_vertices, directed_graph.contains_vertex(num)
+            )
+
+    def test_add_vertices(self):
+        random_vertices: list[int] = random.sample(range(101), 20)
+
+        # build empty graphs
+        undirected_graph = GraphAdjacencyList(vertices=[], edges=[], directed=False)
+        directed_graph = GraphAdjacencyList(vertices=[], edges=[], directed=True)
+
+        # run add_vertex
+        for num in random_vertices:
+            undirected_graph.add_vertex(num)
+
+        for num in random_vertices:
+            directed_graph.add_vertex(num)
+
+        # test add_vertex worked
+        for num in random_vertices:
+            self.__assert_graph_vertex_exists_check(
+                undirected_graph, directed_graph, num
+            )
+
+    def test_remove_vertices(self):
+        random_vertices: list[int] = random.sample(range(101), 20)
+
+        # build graphs WITHOUT edges
+        undirected_graph = GraphAdjacencyList(
+            vertices=random_vertices, edges=[], directed=False
+        )
+        directed_graph = GraphAdjacencyList(
+            vertices=random_vertices, edges=[], directed=True
+        )
+
+        # test remove_vertex worked
+        for num in random_vertices:
+            self.__assert_graph_vertex_exists_check(
+                undirected_graph, directed_graph, num
+            )
+
+            undirected_graph.remove_vertex(num)
+            directed_graph.remove_vertex(num)
+
+            self.__assert_graph_vertex_does_not_exist_check(
+                undirected_graph, directed_graph, num
+            )
+
+    def test_add_and_remove_vertices_repeatedly(self):
+        random_vertices1: list[int] = random.sample(range(51), 20)
+        random_vertices2: list[int] = random.sample(range(51, 101), 20)
+
+        # build graphs WITHOUT edges
+        undirected_graph = GraphAdjacencyList(
+            vertices=random_vertices1, edges=[], directed=False
+        )
+        directed_graph = GraphAdjacencyList(
+            vertices=random_vertices1, edges=[], directed=True
+        )
+
+        # test adding and removing vertices
+        for i in range(len(random_vertices1)):
```

Comment:
There are eight instances of `range(len())` which is almost always a sign that [`enumerate()`](https://docs.python.org/3/library/functions.html#enumerate) should have been used instead.

Reply:
Understood!

Reply:
Actually, I've taken a closer look and while there are a lot of range(len())'s, I believe enumerate will result in less readable code because the index is sometimes used to get values from two separate lists. Let me rewrite it so that `range(len())` is replaced with `range(count)` for optimization.